python/tvm/relay/frontend/pytorch.py

```diff
@@ -955,7 +1025,100 @@ def parse_params(graph, state_dict):
     return params, param_tensors
 
 
-def parse_operators(operators, outputs, output_index_map, ret_name):
+def convert_block(block, outputs, output_index_map):
+    """ Translate Torch "Block", used for prim::If and prim::Loop """
+    ops = _get_operator_nodes(block.nodes())
+    ret_names = _get_input_names(block.returnNode())
+    return convert_operators(ops, outputs, output_index_map, ret_names)
+
+
+def convert_if(if_node, outputs, output_index_map):
+    """ Translate Torch prim::If to Relay If """
+    cond = outputs[output_index_map[if_node.inputsAt(0).debugName()]]
+    blocks = list(if_node.blocks())
+    true_branch = convert_block(blocks[0], outputs, output_index_map)
+    false_branch = convert_block(blocks[1], outputs, output_index_map)
+    assert len(true_branch) == 1 and len(false_branch) == 1
+    return _expr.If(cond, true_branch[0], false_branch[0])
+
+
+def convert_loop(loop_node, outputs, output_index_map):
+    """ Translate Torch prim::Loop to Relay while_loop """
+    def get_input(index):
+        ivalue = loop_node.inputsAt(index)
+        inode = ivalue.node()
+        if inode.kind() == "prim::Constant":
+            return _expr.const(_get_constant(inode))
+        var_name = ivalue.debugName()
+        assert var_name in output_index_map
+        return _wrap_const(outputs[output_index_map[var_name]])
+
+    # Refer to the spec for prim::Loop below
+    # https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#loops
+    # The first input: %max_trip_count
+    # The second input: %initial_condition
+    # The rest of input: loop variables
+    max_loop_count = get_input(0)
+    init_cond = get_input(1)
+    num_loop_var = len(list(loop_node.inputs())) - 2
+    init_vals = [get_input(i + 2) for i in range(num_loop_var)]
+
+    # For loop (not while loop) has always %initial_condition being 1
+    is_for_loop = isinstance(init_cond, _expr.Constant)
```

Comment:
I am not very sure, but I wonder if this is the best way to check if it is for or loop? It it seems from the official tutorial that the difference is:

> Note that translations of for loops simply pass in a constant true for both %initial_condition and %iter_condition, while for while loops %max_trip_count is set to the largest value of int64_t, and %i is unused. Those patterns are recognized by our interpreter and optimized accordingly (e.g. while loops don't maintain the loop counter).

Then, it looks to me that both case could possibly have `isinstance(init_cond, _expr.Constant) == True`, right?

Reply:
thanks for pointing this out, yes, it seems if the condition in in the while loop is not input dependent, such as `while i < 10:`, then init_cond seems to be constant (also see my example IR for while loop above, which shows this). `is_for_loop` becomes True and it is broken.

 For dynamic condition such as `while i < inp.size(0):`, init_cond is not constant. My test cases cover only such cases. I'll add while loop test with constant init cond.

Reply:
@zhiics I added a fix for this in https://github.com/apache/incubator-tvm/pull/4964/commits/cdfc68f12519dce1224f5d3eead7a6f76239431e
can you take a look?

Reply:
Thanks. But technically, it is still possible that `max_trip_count` of a `for_loop` is `sys.maxsize`, right? I agree this should be very rare but I am not sure if this is good enough. It looks that Torch used `%i` to determine if it is a `for` or `while`. `%i` is not used in `while`.

Reply:
I think the fact that `%i` is not used in while case is an implementation detail of torch interpreter. Since iter var increment is not explicit in torch, even for `for` loop case `%i` has no user if we only look at the graph. See below

```
graph(%self : __torch__.SimpleLoop,
      %inp.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %3 : bool = prim::Constant[value=1]() # test.py:9:8
  %4 : int = prim::Constant[value=10]() # test.py:9:23
  %a : Tensor = prim::Loop(%4, %3, %inp.1) # test.py:9:8
    block0(%i.1 : int, %a.5 : Tensor):
      %8 : float = aten::Float(%i.1) # test.py:10:17
      %a.2 : Tensor = aten::add_(%a.5, %8, %2) # test.py:10:12
      -> (%3, %a.2)
  return (%a)
```

Reply:
Thanks for clarification. Let's keep the current implementation for now.